docs/design/2022-06-22-foreign-key.md

```diff
@@ -0,0 +1,762 @@
+
+1. Pseudocode in build plan stage.
+
+```go
+type Delete struct {
+    ...
+    // table-id => foreign key triggers.
+    FKTriggers map[int64][]*ForeignKeyTrigger
+}
+
+type ForeignKeyTrigger struct {
+	Tp                    FKTriggerType
+	OnModifyReferredTable *OnModifyReferredTableFKInfo
+	OnModifyChildTable    *OnModifyChildTableFKInfo
+}
+
+type OnModifyReferredTableFKInfo struct {
+	ReferredFK *model.ReferredFKInfo
+	ChildTable table.Table
+	FK         *model.FKInfo
+}
+
+const (
+	FKTriggerOnDeleteReferredTable      FKTriggerType = 1
+	FKTriggerOnUpdateReferredTable      FKTriggerType = 2
+	FKTriggerOnInsertOrUpdateChildTable FKTriggerType = 3
+)
+
+func (b *PlanBuilder) buildDelete(ctx context.Context, ds *ast.DeleteStmt) (Plan, error) {
+    ....
+    del.FKTriggers = buildOnDeleteForeignKeyTrigger()
+}
+```
+
+2. Pseudocode in build executor.
+
+```go
+type DeleteExec struct {
+    fkTriggerExecs map[int64][]*ForeignKeyTriggerExec
+}
+
+type ForeignKeyTriggerExec struct {
+    fkTrigger     *plannercore.ForeignKeyTrigger
+    colsOffsets   []int
+    fkValues      [][]types.Datum
+}
+
+func (b *executorBuilder) buildDelete(v *plannercore.Delete) Executor {
+	...
+	deleteExec.fkTriggerExecs = b.buildTblID2ForeignKeyTriggerExecs()
+}
+```
+
+3. Pseudocode in `Delete` execution.
+
+```go
+func (e *DeleteExec) removeRow(...){
+    fkTriggerExecs := e.fkTriggerExecs[t.Meta().ID]
+    sc := ctx.GetSessionVars().StmtCtx
+    for _, fkt := range fkTriggerExecs {
+        err = fkt.addRowNeedToTrigger(sc, row)
+    }
+}
+
+func (fkt *ForeignKeyTriggerExec) addRowNeedToTrigger(sc *stmtctx.StatementContext, row []types.Datum) error {
+    vals, err := fetchFKValues(row, fkt.colsOffsets)
+    if err != nil || hasNullValue(vals) {
+        return err
+    }
+    fkt.fkValues = append(fkt.fkValues, vals)
+    return nil
+}
+```
+
+4. Pseudocode after `Delete` execution.
+
+```go
+func (a *ExecStmt) Exec(){
+    e, err := a.buildExecutor()
+    e.Open()
+	f handled, result, err := a.handleNoDelay(ctx, e, isPessimistic); handled {
+        err = a.handleForeignKeyTrigger(ctx, e, isPessimistic)
+        return result, err
+    }
+}
+
+func (a *ExecStmt) handleForeignKeyTrigger(ctx context.Context, e Executor, isPessimistic bool) error {
+    fkTriggerExecs := e.GetForeignKeyTriggerExecs()
+    for i := 0; i < len(fkTriggerExecs); i++{
+        fkt := fkTriggerExecs[i]
+        e, err := fkt.buildExecutor(ctx)
+        if err != nil {
+            return err
+        }
+        if e == nil {
+            continue
+        }
+        if err := e.Open(ctx); err != nil {
+            terror.Call(e.Close)
+            return err
+        }
+        _, _, err = a.handleNoDelay(ctx, e, isPessimistic)
+        if err != nil {
+            return err
+        }
+        err = a.handleForeignKeyTrigger(ctx, e, isPessimistic)
+        if err != nil {
+            return err
+        }
+    }
+    return nil
+}
+```
+
+### Issue need to be discussed
```

Comment:
Do we need to add https://github.com/pingcap/tidb/pull/37148#discussion_r957951003 to here?

Reply:
done.